sled-agent/src/storage_manager.rs

```diff
@@ -105,131 +116,379 @@ impl Pool {
+            .join(self.id.to_string());
+        create_dir_all(&path).await?;
+        let mut path = path.join(dataset_id.to_string());
+        path.set_extension("toml");
+        Ok(path)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, JsonSchema, Clone)]
+struct DatasetName {
+    // A unique identifier for the Zpool on which the dataset is stored.
+    pool_name: String,
+    // A name for the dataset within the Zpool.
+    dataset_name: String,
+}
+
+impl DatasetName {
+    fn new(pool_name: &str, dataset_name: &str) -> Self {
+        Self {
+            pool_name: pool_name.to_string(),
+            dataset_name: dataset_name.to_string(),
+        }
+    }
+
+    fn full(&self) -> String {
+        format!("{}/{}", self.pool_name, self.dataset_name)
+    }
 }
 
 // Description of a dataset within a ZFS pool, which should be created
 // by the Sled Agent.
-struct PartitionInfo<'a> {
-    name: &'a str,
-    zone_prefix: &'a str,
-    data_directory: &'a str,
-    svc_directory: &'a str,
-    port: u16,
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+struct DatasetInfo {
+    // TODO: Is this always "/data"?
+    data_directory: String,
+    address: SocketAddr,
     kind: DatasetKind,
+    name: DatasetName,
 }
 
+impl DatasetInfo {
+    fn new(pool: &str, kind: DatasetKind, address: SocketAddr) -> DatasetInfo {
+        match kind {
+            DatasetKind::CockroachDb { .. } => DatasetInfo {
+                name: DatasetName::new(pool, "cockroachdb"),
+                data_directory: "/data".to_string(),
+                address,
+                kind,
+            },
+            DatasetKind::Clickhouse { .. } => DatasetInfo {
+                name: DatasetName::new(pool, "clickhouse"),
+                data_directory: "/data".to_string(),
+                address,
+                kind,
+            },
+            DatasetKind::Crucible { .. } => DatasetInfo {
+                name: DatasetName::new(pool, "crucible"),
+                data_directory: "/data".to_string(),
+                address,
+                kind,
+            },
+        }
+    }
+
+    fn zone_prefix(&self) -> String {
+        format!("{}{}_", ZONE_PREFIX, self.name.full())
+    }
+
+    async fn start_zone(
+        &self,
+        log: &Logger,
+        zone: &RunningZone,
+        address: SocketAddr,
+        do_format: bool,
+    ) -> Result<(), Error> {
+        match self.kind {
+            DatasetKind::CockroachDb { .. } => {
+                // Load the CRDB manifest.
+                zone.run_cmd(&[
+                    crate::illumos::zone::SVCCFG,
+                    "import",
+                    "/var/svc/manifest/site/cockroachdb/manifest.xml",
+                ])?;
+
+                // Set parameters which are passed to the CRDB binary.
+                zone.run_cmd(&[
+                    crate::illumos::zone::SVCCFG,
+                    "-s",
+                    "svc:system/illumos/cockroachdb",
```

Comment:
`svc:system/illumos/` could be pulled into a constant, or configurable

Reply:
I'm pulling it into constants at the top of this file.

There is still a *tiny* amount of duplication, because `concat!` doesn't appear to work well with tokens (even constant ones) without using some external crates.